apps/state/lib/state/alert/hooks.ex

```diff
@@ -17,9 +17,60 @@ defmodule State.Alert.Hooks do
     entities
     |> Stream.flat_map(&include_entity_parent_stop/1)
     |> Stream.flat_map(&include_entity_alternate_trips/1)
+    |> Enum.group_by(&get_key/1)
+    |> Stream.map(&merge_entities/1)
     |> Enum.uniq()
   end
 
+  defp get_key(%{} = informed_entity) do
+    %{
+      route: Map.get(informed_entity, :route),
+      stop: Map.get(informed_entity, :stop),
+      trip: Map.get(informed_entity, :trip)
+    }
+  end
+
+  defp merge_entities({%{} = key, entities}) do
```

Comment:
Well, this probably would work as well but I've tried to implement it and it doesn't appear to be any _simpler_--neither in terms of complexity nor in terms of LoC--than the current implementation, which I've already thoroughly tested. Therefore, unless you see serious flaws in the proposed approach, I'd prefer to stick with it.

Reply:
It seems like there are more functions here than I would have expected, rather than something like:

```elixir
defp merge_entities({key, entities}) do
  activities = merge_activities(entities)
  if MapSet.empty?(activities) do
    entities
  else
    for entity <- entities, do: Map.put(entity, :activities, activities)
  end
end
```

Reply:
Refactored.